### PR TITLE
Fix verious GL errors in GLES 3.1 mode and Mario tennis.

### PR DESCRIPTION
**Pull request by fzurita**
This fixes all the GL errors I encountered in GLES 3.1 and Mario Tennis. See https://github.com/gonetz/GLideN64/issues/1110.

### CODE REVIEW

**Comment by fzurita** [comment]
Also, thanks to the Android GL Errors traceability, I was able to find why Multipsampling wasn't working, at least in Android. I added that to the pull request.

File: src/OGL3X/GLSLCombiner_ogl3x.cpp [hunk 1]
```diff
@@ -142,18 +142,25 @@ void InitZlutTexture()
...
 	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+#ifndef GLESX
 	glTexImage2D(GL_TEXTURE_2D, 0, fboFormats.lutInternalFormat,
 		512, 512, 0, fboFormats.lutFormat, fboFormats.lutType,
 		zLUT);
-	glBindImageTexture(ZlutImageUnit, g_zlut_tex, 0, GL_FALSE, 0, GL_READ_ONLY, fboFormats.lutInternalFormat);
+#else
+	glTexStorage2D(	GL_TEXTURE_2D, 1, fboFormats.lutInternalFormat, 512, 512);
+#endif
```
Comment (gonetz) [inline src/OGL3X/GLSLCombiner_ogl3x.cpp]: Please explain that part. Original code uses glTexImage2D with zLUT as texture data. Why you change that and when g_zlut_tex will be filled with data?

Comment (fzurita) [inline src/OGL3X/GLSLCombiner_ogl3x.cpp]: glBindImageTexture requires an immutable texture object. See https://www.khronos.org/opengles/sdk/docs/man31/html/glBindImageTexture.xhtml

glTexStorage2D doesn't create immutable texture objects but glTexStorage2D does.

I think you did make me realize a mistake. glBindImageTexture takes texture data as an argument, glTexStorage2D doesn't. I need to find how to set texture data after.

Comment (fzurita) [inline src/OGL3X/GLSLCombiner_ogl3x.cpp]: I think I can use glTexSubImage2D to update the data after.

**Comment by fzurita** [comment]
I updated the pull request with a fix for the error you made me realize.

File: src/OGL3X/GLSLCombiner_ogl3x.cpp [hunk 2]
```diff
@@ -142,18 +142,26 @@ void InitZlutTexture()
...
 	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+#ifndef GLESX
 	glTexImage2D(GL_TEXTURE_2D, 0, fboFormats.lutInternalFormat,
 		512, 512, 0, fboFormats.lutFormat, fboFormats.lutType,
 		zLUT);
-	glBindImageTexture(ZlutImageUnit, g_zlut_tex, 0, GL_FALSE, 0, GL_READ_ONLY, fboFormats.lutInternalFormat);
+#else
+	glTexStorage2D(	GL_TEXTURE_2D, 1, fboFormats.lutInternalFormat, 512, 512);
+	glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, 512, 512, fboFormats.lutFormat, fboFormats.lutType, zLUT);
```
Comment (gonetz) [inline src/OGL3X/GLSLCombiner_ogl3x.cpp]: This must work with GL as well, so let make that code common. Please replace the old code, without ifdef endif

Comment (fzurita) [inline src/OGL3X/GLSLCombiner_ogl3x.cpp]: Ok, will do.

Comment (fzurita) [inline src/OGL3X/GLSLCombiner_ogl3x.cpp]: Ok, done.

**Comment by gonetz** [comment]
Thanks! Merged: 010c415d7186

**Comment by gonetz** [comment]
OOPS! Assert in OpenGl.cpp line 418
assert(m_textureBoundsLoc >= 0);

**Comment by fzurita** [comment]
Hmm, not good, I should probably start running debug soi that I will hit the asserts. Any idea what's causing that? The changes I made should had been benign.

**Comment by gonetz** [comment]
The code looks correct. I don't see any problems, but glsl compiler removes the uniform now. I can't find how to fix it.

**Comment by fzurita** [comment]
Yikes, it sounds like a GLSL compiler bug, it must be removing the variable because it thinks that it's not being used, but both versions of the macro is clearly using it.

You could make it easier for the GLSL compiler, in the main method of the shader, you could do:
```
mediump vec4 temp = uTextureBounds;
```

as the first call. Then, from then on, instead of using ```uTextureBounds```, use ```temp```.

**Comment by gonetz** [comment]
Does not help;

**Comment by fzurita** [comment]
Is the compilation of the shader succeeding, but m_textureBoundsLoc is not being located?

**Comment by gonetz** [comment]
Yes

**Comment by gonetz** [comment]
```
const char * strTexrectDrawerTexBilinearFilter =
MAIN_SHADER_VERSION
"#define TEX_OFFSET(off, tex, texCoord, texSize) texture(tex, texCoord - (off)/texSize)							\n"
"#define TEX_FILTER(name, tex, texCoord)															\\\n"
"{																												\\\n"
"  mediump vec2 texSize = vec2(textureSize(tex,0));																\\\n"
"  mediump vec2 texelSize = vec2(1.0) / texSize;																\\\n"
"  lowp vec4 c = texture(tex, texCoord);																		\\\n"
"  if (abs(texCoord.s - textureBounds[0]) < texelSize.x || abs(texCoord.s - textureBounds[2]) < texelSize.x){	\\\n"
"    name = c;												 													\\\n"
"  } else if (abs(texCoord.t - textureBounds[1]) < texelSize.y || abs(texCoord.t - textureBounds[3]) < texelSize.y){	\\\n"
"    name = c;												 													\\\n"
"  } else {													 													
...
"																												\n"
;

const char * strTexrectDrawerFragmentShaderTex =
"uniform sampler2D uTex0;																						\n"
"uniform lowp int uEnableAlphaTest;																				\n"
"uniform mediump vec4 uTextureBounds;																			\n"
"lowp vec4 uTestColor = vec4(4.0/255.0, 2.0/255.0, 1.0/255.0, 0.0);												\n"
"in mediump vec2 vTexCoord0;																					\n"
"out lowp vec4 fragColor;																						\n"
"void main()																									\n"
"{																												\n"
" mediump vec4 textureBounds = uTextureBounds;																			"  TEX_FILTER(fragColor, uTex0, vTexCoord0);													\n"
"  if (fragColor == uTestColor) discard;																		\n"
"  if (uEnableAlphaTest == 1 && !(fragColor.a > 0.0)) discard;													\n"
"}																												\n"
;

```

**Comment by gonetz** [comment]
Definitely compiler error: when I replace "name = c;" by "discard;" the problem gone. Looks like the compiler can't detect that "name = c;" affect shader output.

**Comment by gonetz** [comment]
You also did a mistake in this macro:
```
"    name = mix( pInterp_q0, pInterp_q1, interpolationFactor.y ); 												\\\n" // Interpolate in Y direction.
"    name = c;												 													\\\n"

```
I overlooked it.

**Comment by fzurita** [comment]
Whoops, I missed that.

**Comment by gonetz** [comment]
Yep, that was the reason! When I removed wrong line
`"    name = c;"`
compiler found the variable. It was not compiler error. uTextureBound is redundant if in all cases name = c.

**Comment by fzurita** [comment]
Yep, a case of the compiler being too smart!

**Comment by gonetz** [comment]
Just as smart as it should be. Otherwise we would not find that bug.

**Comment by fzurita** [comment]
The Adreno compiler wasn't as smart unfortunately, otherwise I would have had the same issue.

**Comment by gonetz** [comment]
3point version of the shader has compilation error
```
Fragment shader failed to compile with the following errors:
ERROR: 0:31: error(#112) Too few args in Macro: TEX_FILTER
ERROR: 0:31: error(#202) No matching overloaded function found: textureSize
ERROR: 0:31: error(#132) Syntax error: ")" parse error
ERROR: error(#273) 3 compilation errors.  No code generated

```

**Comment by gonetz** [comment]
Sorry, it is probably my fault :(

**Comment by gonetz** [comment]
Yes, I did not remove some debug modifications. It is ok now. I'll push the patch.

**Comment by gonetz** [comment]
Done.

**Comment by gonetz** [comment]
Re-done.

**Comment by fzurita** [comment]
Cool, thanks. Sorry about that bad code in there.

**Pull request closed without merging**